### PR TITLE
[CodeCamp2023-500]add large_image_demo

### PR DESCRIPTION
**Pull request by YanxingLiu**
## Motivation

To add large image demo in mmdetection.

## Modification

demo/large_image.jpg
demo/large_image_demo.py
mmdet/utils/large_image.py
mmdet/utils/misc.py

## Use cases (Optional)

```shell
wget -P checkpoint https://download.openmmlab.com/mmdetection/v2.0/faster_rcnn/faster_rcnn_r101_fpn_2x_coco/faster_rcnn_r101_fpn_2x_coco_bbox_mAP-0.398_20200504_210455-1d2dac9c.pth # noqa: E501, E261.

python demo/large_image_demo.py \
    demo/large_image.jpg \
    configs/faster_rcnn/faster-rcnn_r101_fpn_2x_coco.py \
    checkpoint/faster_rcnn_r101_fpn_2x_coco_bbox_mAP-0.398_20200504_210455-1d2dac9c.pth
```
result:
image

# parameter test:
* cpu inference: pass
* --show: pass
image
* --tta: pass
```shell
wget -P checkpoint https://download.openmmlab.com/mmdetection/v2.0/retinanet/retinanet_r50_fpn_1x_coco/retinanet_r50_fpn_1x_coco_20200130-c2398f9e.pth # noqa: E501, E261.

python demo/large_image_demo.py \
    demo/large_image.jpg \
    configs/retinanet/retinanet_r50_fpn_1x_coco.py \
    checkpoint/retinanet_r50_fpn_1x_coco_20200130-c2398f9e.pth --tta
```
result:
image
* --score-thr: pass
* --path-size: pass
* --patch-overlap-ratio: pass
* --merge-iou-thr: pass
* --merge-nms-type: pass
* --batch-size: pass
* --debug: pass
* --save-patch: pass

### CODE REVIEW

**Comment by hhaAndroid** [comment]
@YanxingLiu Is it possible to support instance segmentation?

**Comment by YanxingLiu** [comment]
> @YanxingLiu Is it possible to support instance segmentation?

I am not familiar with instance segmentation, bu  it feels like it might be more difficult than object detection.

**Comment by hhaAndroid** [comment]
> > @YanxingLiu Is it possible to support instance segmentation?
> 
> I am not familiar with instance segmentation, but it feels like it might be more difficult than object detection.

OK

File: demo/large_image_demo.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        assert 'tta_pipeline' in config, 'Cannot find ``tta_pipeline`` ' \
+                                         "in config. Can't use tta !"
+        config.model = ConfigDict(**config.tta_model, module=config.model)
+        test_data_cfg = config.test_dataloader.dataset
+        while 'dataset' in test_data_cfg:
+            test_data_cfg = test_data_cfg['dataset']
+
+        # batch_shapes_cfg will force control the size of the output image,
+        # it is not compatible with tta.
+        if 'batch_shapes_cfg' in test_data_cfg:
```
Comment (hhaAndroid) [inline demo/large_image_demo.py:117]: delete

File: demo/large_image_demo.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (hhaAndroid) [inline demo/large_image_demo.py:1]: Please refer to the mmyolo documentation and add relevant instructions to the Chinese and English documentation in `docs/zh_cn/user_guides/inference.md` in mmdet.

Comment (hhaAndroid) [reply demo/large_image_demo.py:1]: Please make sure that you have tested all the functionalities and include the test results in the PR message. Thank you.

File: demo/large_image_demo.py (hunk 1 above)
Comment (YanxingLiu) [reply demo/large_image_demo.py]: ok, I have deteled it.

File: demo/large_image_demo.py (hunk 2 above)
Comment (YanxingLiu) [reply demo/large_image_demo.py:1]: > Please make sure that you have tested all the functionalities and include the test results in the PR message. Thank you.

I have tested all the arguments in large_image_demo, and some necessary results are shown in the PR message. Thank you.

Comment (YanxingLiu) [reply demo/large_image_demo.py:1]: > Please refer to the mmyolo documentation and add relevant instructions to the Chinese and English documentation in `docs/zh_cn/user_guides/inference.md` in mmdet.

I have added Chinese and English version in docs.

**Review by hhaAndroid — APPROVED** [review]

**Comment by W-hary** [comment]
When I run 'large_image_demo.py' for object detection on 'large image.jpg', I can successfully get the detection results. But when I use 'rtmdet-ins_tiny_8xb32-300e_coco.py' to instance split 'large image.jpg', an error will be reported: AssertionError: `binary_masks` must have the same shape with image
Does 'large_image_demo.py' not work for instance split testing? How can I solve this problem
59d93b1a0859b8a12501844d1832331

**Comment by DAIJinJ** [comment]
> When I run 'large_image_demo.py' for object detection on 'large image.jpg', I can successfully get the detection results. But when I use 'rtmdet-ins_tiny_8xb32-300e_coco.py' to instance split 'large image.jpg', an error will be reported: AssertionError: `binary_masks` must have the same shape with image Does 'large_image_demo.py' not work for instance split testing? How can I solve this problem 59d93b1a0859b8a12501844d1832331

same problem

**Pull request merged**